loki/transformations/remove_code.py

```diff
@@ -85,30 +90,86 @@ def __init__(
         self.remove_imports = remove_imports
 
         self.kernel_only = kernel_only
+        self.remove_unused_args = remove_unused_args
 
     def transform_subroutine(self, routine, **kwargs):
 
-        if self.kernel_only and not kwargs.get('role') == 'kernel':
-            return
+        if kwargs.get('role') == 'kernel' or not self.kernel_only:
+            # Apply named node removal to strip specific calls
+            if self.call_names or self.intrinsic_names:
+                do_remove_calls(
+                    routine, call_names=self.call_names,
+                    intrinsic_names=self.intrinsic_names,
+                    remove_imports=self.remove_imports
+                )
+
+            # Apply marked region removal
+            if self.remove_marked_regions:
+                do_remove_marked_regions(
+                    routine, mark_with_comment=self.mark_with_comment
+                )
+
+            # Apply Dead Code Elimination
+            if self.remove_dead_code:
+                do_remove_dead_code(routine, use_simplify=self.use_simplify)
+
+        if self.remove_unused_args and (item := kwargs['item']):
+            if item.config.get('remove_unused_args', True):
+                successors = kwargs['sub_sgraph'].successors(item=item)
+                self.do_remove_unused_args(routine, item, successors, kwargs['role'])
+
+    def do_remove_unused_args(self, routine, item, successors, role):
+
+        # update callstatements
+        self._update_callstatements(routine, successors)
+
+        if role == 'kernel':
+            # find unused args
+            unused_args = self._find_unused_args(routine)
+
+            # store unused args
+            item.trafo_data[self._key] = {'unused_args': unused_args}
+
+            # remove unused args
+            routine.variables = [a for a in routine.variables
+                                 if not a.name.lower() in unused_args]
+
+    @staticmethod
```

Comment:
I know we used this pattern a lot before, but it's actually not great tying this to a class for no reason. Could we instead simply make this a public utility method of this module? This makes re-use in other modules easier, I think.

And making it "public" (removing the `_`) then also warrants a docstring, I'd suggest. 😉